### PR TITLE
impr: (VM) gas 

### PR DESCRIPTION
**Pull request by jorgemmsilva**
- min gas per request
- deprecated `GasNominalUnit` and `GasPricePerNominalUnit`, gas price now set by: `GasPerToken`
- improved gas estimation
- added test for deposits/withdrawals edge cases

### CODE REVIEW

File: packages/vm/gas/table.go [hunk 1]
```diff
@@ -59,6 +60,7 @@ var burnTable = BurnTable{
 	BurnCodeUtilsBLSValidSignature:     {"bls valid", constValue(2000)},
 	BurnCodeUtilsBLSAddrFromPubKey:     {"bls addr", constValue(50)},
 	BurnCodeUtilsBLSAggregateBLS1P:     {"bls aggregate", linear(CoefBLSAggregate)},
+	BurnCodeMinimumGasPerRequest:       {"minimum gas per request", constValue(100)}, // TODO maby make it configurable (gov contract?)
```
Comment (dessaya) [inline packages/vm/gas/table.go]:
```suggestion
	BurnCodeMinimumGasPerRequest:       {"minimum gas per request", constValue(100)}, // TODO maybe make it configurable (gov contract?)
```

File: packages/vm/gas/policy.go [hunk 2]
```diff
@@ -11,69 +11,58 @@ import (
...
 }
 
 func DefaultGasFeePolicy() *GasFeePolicy {
 	return &GasFeePolicy{
-		GasFeeTokenID:          nil, // default is iotas
-		GasNominalUnit:         100, // gas is burned in 100-s and not less than 100
-		GasPricePerNominalUnit: 1,   // default is 1 iota = 100 gas
-		ValidatorFeeShare:      0,   // by default all goes to the governor
+		GasFeeTokenID:     nil, // default is iotas
+		GasPerToken:       100, // gas is burned in 100-s and not less than 100
```
Comment (dessaya) [inline packages/vm/gas/policy.go]: `gas is burned in 100-s and not less than 100`
Is this still true?

**Review by dessaya — APPROVED** [review]

File: packages/vm/gas/policy.go (hunk 2 above)
Comment (jorgemmsilva) [reply packages/vm/gas/policy.go]: I mean with `GasPerToken = 100` you would need 2 tokens to burn `101` gas, so each tokens would entitle you to 100 gas units. But the wording is kinda confusing, will edit 👍

File: packages/vm/gas/policy.go [hunk 3]
```diff
@@ -11,68 +11,48 @@ import (
 type GasFeePolicy struct {
 	// GasFeeTokenID contains iotago.NativeTokenID used to pay for gas, or nil if iotas are used for gas fee
 	GasFeeTokenID *iotago.NativeTokenID
-	// GasNominalUnit price is specified per how many gas units
-	GasNominalUnit uint64
-	// GasPricePer1000Units how many gas tokens you pay for 1000 gas
-	GasPricePerNominalUnit uint64
+	// GasPerToken specifies how many gas units are paid for each token ( 100 means 1 tokens pays for 100 gas)
+	GasPerToken uint64
```
Comment (lunfardo314) [inline packages/vm/gas/policy.go:15]: we are losing option to specify for example `1 token per 1000 gas`? That may be a big limiting factor if we want gas fee to be paid for example by ASM token which has big market value. 
The point of nominals is about:
- being able to choose a token for paying a fee
- being independent from the price of the token

**Review by lunfardo314 — APPROVED** [review]

File: packages/vm/gas/policy.go (hunk 3 above)
Comment (jorgemmsilva) [reply packages/vm/gas/policy.go:15]: You're absolutely correct, but as we talked about in the team call - let's keep it simple for now (accept the current limitation that its not possible to increase the price to more that 1 token = 1 gas unit) and we shall implement a more sophisticated gas pricing when needed :+1:
